pkg/app_gtwy/server/aether-app-gtwy-impl.go

```diff
@@ -0,0 +1,684 @@
+// SPDX-FileCopyrightText: 2020-present Open Networking Foundation <info@opennetworking.org>
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+package server
+
+import (
+	"context"
+	"fmt"
+	"github.com/labstack/echo/v4"
+	externalRef0 "github.com/onosproject/aether-roc-api/pkg/aether_2_0_0/types"
+	"github.com/onosproject/aether-roc-api/pkg/app_gtwy/types"
+	"github.com/onosproject/aether-roc-api/pkg/southbound"
+	"github.com/onosproject/aether-roc-api/pkg/utils"
+	externalRef1 "github.com/onosproject/config-models/modelplugin/aether-2.0.0/aether_2_0_0"
+	"github.com/onosproject/onos-lib-go/pkg/logging"
+	"github.com/openconfig/gnmi/proto/gnmi"
+	"github.com/prometheus/common/model"
+	"net/http"
+	"strconv"
+	"time"
+)
+
+var log = logging.GetLogger("app_gtwy")
+
+// AppGtwy -
+type AppGtwy struct {
+	GnmiClient      southbound.GnmiClient
+	GnmiTimeout     time.Duration
+	AnalyticsClient AnalyticsClient
+}
+
+// GetDevices /appgtwy/v1/{target}/enterprises/{enterprise-id}/sites/{site-id}/devices
+func (i *AppGtwy) GetDevices(ctx echo.Context, target externalRef0.Target, enterpriseID string, siteID string) error {
+
+	var listDeviceResponse []types.AppGtwyDevice
+
+	gnmiCtx, cancel := utils.NewGnmiContext(ctx, i.GnmiTimeout)
+	defer cancel()
+
+	site, err := i.gnmiGetEnterprisesEnterpriseSite(gnmiCtx, "/aether/v2.0.0/{target}/enterprises/enterprise/{enterprise-id}/site/{site-id}", target, enterpriseID, siteID)
+	if err != nil {
+		return utils.ConvertGrpcError(err)
+	}
+
+	si, err := i.AnalyticsClient.Query("subscriber_info")
+	if err != nil {
+		log.Errorf("error querying prometheus", err)
+		return err
+	}
+
+	simMap := mapSim(site)
+	deviceGroupMap := mapDeviceToDeviceGroups(site)
+	imsiMap := mapIMSI(si)
+
+	deviceMap := make(map[string]types.AppGtwyDevice)
+	for _, d := range *site.Device {
+		s := simMap[*d.SimCard]
+		newDev := types.AppGtwyDevice{
+			DeviceId:     &d.DeviceId,
+			DisplayName:  d.DisplayName,
+			Description:  d.Description,
+			Imei:         d.Imei,
+			SimIccid:     s.Iccid,
+			DeviceGroups: deviceGroupMap[d.DeviceId],
+		}
+		deviceMap[d.DeviceId] = newDev
+
+		imsi := strconv.Itoa(int(*s.Imsi))
+		if _, ok := imsiMap[imsi]; ok {
+			ip := string(imsiMap[imsi].Metric["ip"])
+			newDev.Ip = &ip
+			attached := imsiMap[imsi].Value.String()
+			newDev.Attached = &attached
+		}
+
+		listDeviceResponse = append(listDeviceResponse, newDev)
+	}
+
+	log.Infof("GetAppGtwyDevices")
```

Comment:
I think it's always good to include some dynamic info in a log message - e.g. even the number of devices would be a help